Plot options rewrite

kecnry opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request implements custom jdaviz plot options that we can modify/style as well as an "advanced" mode (off by default) that allows for editing plot options across multiple viewers/layer simultaneously.

This PR does not address all the styling changes yet, but provides the framework for us to make those changes in follow-up efforts.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->


### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

Jenneh reviewed (approved):
Looks good. I see more styling updates for the general plugin but that is outside the scope of this ticket.

pllim commented:
Needs a rebase to fix change log -- FYI

pllim reviewed (changes requested):
I get this traceback if I create a second viewer in Imviz, load an image in the second viewer, and then select the second viewer in Plot Options.

```
.../ipywidgets/widgets/widget.py in _handle_msg(self, msg)
    674                 if 'buffer_paths' in data:
    675                     _put_buffers(state, data['buffer_paths'], msg['buffers'])
--> 676                 self.set_state(state)
    677 
    678         # Handle a state request.

.../ipywidgets/widgets/widget.py in set_state(self, sync_data)
    543                     from_json = self.trait_metadata(name, 'from_json',
    544                                                     self._trait_from_json)
--> 545                     self.set_trait(name, from_json(sync_data[name], self))
    546 
    547     def send(self, content, buffers=None):

.../contextlib.py in __exit__(self, typ, value, traceback)
    124         if typ is None:
    125             try:
--> 126                 next(self.gen)
    127             except StopIteration:
    128                 return False

.../traitlets/traitlets.py in hold_trait_notifications(self)
   1214                 for changes in cache.values():
   1215                     for change in changes:
-> 1216                         self.notify_change(change)
   1217 
   1218     def _notify_trait(self, name, old_value, new_value):

.../ipywidgets/widgets/widget.py in notify_change(self, change)
    604                 # Send new state to front-end
    605                 self.send_state(key=name)
--> 606         super(Widget, self).notify_change(change)
    607 
    608     def __repr__(self):

.../traitlets/traitlets.py in notify_change(self, change)
   1227     def notify_change(self, change):
   1228         """Notify observers of a change event"""
-> 1229         return self._notify_observers(change)
   1230 
   1231     def _notify_observers(self, event):

.../traitlets/traitlets.py in _notify_observers(self, event)
   1264                 c = getattr(self, c.name)
   1265 
-> 1266             c(event)
   1267 
   1268     def _add_notifiers(self, handler, name, type):

.../jdaviz/core/template_mixin.py in _on_viewer_changed(self, msg)
    448         if new != old:
    449             self._clear_cache()
--> 450             self._on_layers_changed()
    451 
    452     @observe('filters')

.../jdaviz/core/template_mixin.py in _on_layers_changed(self, msg)
    457         if not isinstance(viewer_names, list):
    458             viewer_names = [viewer_names]
--> 459         viewers = [self.app.get_viewer(viewer) for viewer in viewer_names]
    460 
    461         manual_items = [{'label': label} for label in self.manual_options]

.../jdaviz/core/template_mixin.py in <listcomp>(.0)
    457         if not isinstance(viewer_names, list):
    458             viewer_names = [viewer_names]
--> 459         viewers = [self.app.get_viewer(viewer) for viewer in viewer_names]
    460 
    461         manual_items = [{'label': label} for label in self.manual_options]

.../jdaviz/app.py in get_viewer(self, viewer_reference)
    439             The viewer class instance.
    440         """
--> 441         return self._viewer_by_reference(viewer_reference)
    442 
    443     def get_viewer_by_id(self, vid):

.../jdaviz/app.py in _viewer_by_reference(self, reference)
    997         viewer_item = self._viewer_item_by_reference(reference)
    998 
--> 999         return self._viewer_store[viewer_item['id']]
   1000 
   1001     def _viewer_item_by_reference(self, reference):

TypeError: 'NoneType' object is not subscriptable
```

pllim commented:
I would argue that maybe we don't even need "Show axes" for Imviz. If you enable it and then disable it again, there is whitespace around the viewer that you cannot get rid of, reducing the display real estate.

![Screenshot 2022-05-25 165639](https://user-images.githubusercontent.com/2090236/170365824-2a7bfac6-c9a1-4574-bd89-e7c98ae3ff63.png)

kecnry commented:
Rebased and fixed handling retrieving layers for imviz viewers without names.

> I would argue that maybe we don't even need "Show axes" for Imviz. If you enable it and then disable it again, there is whitespace around the viewer that you cannot get rid of, reducing the display real estate.

It's easy to hide if that's what we want to do (maybe there are cases where you want to see the scale, but at least this information is available from toolbar when mousing over pixels).  The whitespace is an issue upstream that also affects cubeviz/mosviz - we need to force a redraw when toggling the axes.  If you resize the viewer panel, it will force it to adjust and fill the space appropriately.  I'll bring this up for discussion and we can decide whether to keep or remove the toggle in imviz.

pllim commented:
Fails in a different way now. 😬 

![Screenshot 2022-05-26 103713](https://user-images.githubusercontent.com/2090236/170510604-d99df350-e93c-404a-9b0d-d17d4873aeba.png)

```
.../ipyvue/VueTemplateWidget.py in _handle_event(self, _, content, buffers)
     58                 getattr(self, "vue_" + event)(data, buffers)
     59             else:
---> 60                 getattr(self, "vue_" + event)(data)
     61 
     62 

.../jdaviz/configs/imviz/plugins/image_viewer_creator/image_viewer_creator.py in vue_create_image_viewer(self, *args, **kwargs)
     20             ImvizImageView, data=None, sender=self)
     21 
---> 22         self.hub.broadcast(new_viewer_message)

.../glue/core/hub.py in broadcast(self, message)
    213             logging.getLogger(__name__).info("Broadcasting %s", message)
    214             for subscriber, handler in self._find_handlers(message):
--> 215                 handler(message)
    216 
    217     def __getstate__(self):

.../jdaviz/app.py in _on_new_viewer(self, msg, vid, name)
   1411 
   1412         # Send out a toast message
-> 1413         self.hub.broadcast(ViewerAddedMessage(vid, sender=self))
   1414 
   1415         return viewer

.../glue/core/hub.py in broadcast(self, message)
    213             logging.getLogger(__name__).info("Broadcasting %s", message)
    214             for subscriber, handler in self._find_handlers(message):
--> 215                 handler(message)
    216 
    217     def __getstate__(self):

.../jdaviz/core/template_mixin.py in _on_viewers_changed(self, msg)
    915         self.items = manual_items + [{k: v for k, v in vd.items() if k != 'viewer'}
    916                                      for vd in self.viewer_dicts if self._is_valid_item(vd['viewer'])] # noqa
--> 917         self._apply_default_selection()
    918 
    919 

.../jdaviz/core/template_mixin.py in _apply_default_selection(self)
    310 
    311         if self.default_mode == 'first':
--> 312             self.selected = self.labels[0] if len(self.labels) else ''
    313         elif self.default_mode == 'default_text':
    314             self.selected = self._default_text if self._default_text else ''

.../jdaviz/core/template_mixin.py in __setattr__(self, attr, value, force_super)
    116             return super().__setattr__(attr, value)
    117 
--> 118         return setattr(self._plugin, self._plugin_traitlets.get(attr), value)
    119 
    120     def _clear_cache(self, *attrs):

.../traitlets/traitlets.py in __set__(self, obj, value)
    604             raise TraitError('The "%s" trait is read-only.' % self.name)
    605         else:
--> 606             self.set(obj, value)
    607 
    608     def _validate(self, obj, value):

.../traitlets/traitlets.py in set(self, obj, value)
    593             # we explicitly compare silent to True just in case the equality
    594             # comparison above returns something other than True/False
--> 595             obj._notify_trait(self.name, old_value, new_value)
    596 
    597     def __set__(self, obj, value):

.../traitlets/traitlets.py in _notify_trait(self, name, old_value, new_value)
   1217 
   1218     def _notify_trait(self, name, old_value, new_value):
-> 1219         self.notify_change(Bunch(
   1220             name=name,
   1221             old=old_value,

.../ipywidgets/widgets/widget.py in notify_change(self, change)
    604                 # Send new state to front-end
    605                 self.send_state(key=name)
--> 606         super(Widget, self).notify_change(change)
    607 
    608     def __repr__(self):

.../traitlets/traitlets.py in notify_change(self, change)
   1227     def notify_change(self, change):
   1228         """Notify observers of a change event"""
-> 1229         return self._notify_observers(change)
   1230 
   1231     def _notify_observers(self, event):

.../traitlets/traitlets.py in _notify_observers(self, event)
   1264                 c = getattr(self, c.name)
   1265 
-> 1266             c(event)
   1267 
   1268     def _add_notifiers(self, handler, name, type):

.../jdaviz/core/template_mixin.py in _on_viewer_layer_changed(self, msg)
   1432         current_glue_values = []
   1433         self._linked_states = []
-> 1434         for state, icon in zip(self.subscribed_states, self.subscribed_icons):
   1435             if not self.state_filter(state):
   1436                 continue

.../jdaviz/core/template_mixin.py in __getattr__(self, attr)
    108     def __getattr__(self, attr):
    109         if attr[0] == '_' or attr not in self._plugin_traitlets.keys():
--> 110             return super().__getattribute__(attr)
    111 
    112         return getattr(self._plugin, self._plugin_traitlets.get(attr))

.../functools.py in __get__(self, instance, owner)
    967                 val = cache.get(self.attrname, _NOT_FOUND)
    968                 if val is _NOT_FOUND:
--> 969                     val = self.func(instance)
    970                     try:
    971                         cache[self.attrname] = val

.../jdaviz/core/template_mixin.py in subscribed_states(self)
   1378         # states object that according to the viewer selection, we *should*
   1379         # link if this entry is found there
-> 1380         viewers = self._viewer_select.selected_obj
   1381         if not isinstance(viewers, list):
   1382             # which is the case for single-select

.../jdaviz/core/template_mixin.py in __getattr__(self, attr)
    108     def __getattr__(self, attr):
    109         if attr[0] == '_' or attr not in self._plugin_traitlets.keys():
--> 110             return super().__getattribute__(attr)
    111 
    112         return getattr(self._plugin, self._plugin_traitlets.get(attr))

.../functools.py in __get__(self, instance, owner)
    967                 val = cache.get(self.attrname, _NOT_FOUND)
    968                 if val is _NOT_FOUND:
--> 969                     val = self.func(instance)
    970                     try:
    971                         cache[self.attrname] = val

.../jdaviz/core/template_mixin.py in selected_obj(self)
    875         if self.is_multiselect and len(self.selected):
    876             return [self._get_selected_obj(selected, selected_id)
--> 877                     for selected, selected_id in zip(self.selected, self.selected_id)]
    878         return self._get_selected_obj(self.selected, self.selected_id)
    879 

.../jdaviz/core/template_mixin.py in __getattr__(self, attr)
    108     def __getattr__(self, attr):
    109         if attr[0] == '_' or attr not in self._plugin_traitlets.keys():
--> 110             return super().__getattribute__(attr)
    111 
    112         return getattr(self._plugin, self._plugin_traitlets.get(attr))

.../jdaviz/core/template_mixin.py in selected_id(self)
    860     @property
    861     def selected_id(self):
--> 862         return self.selected_item.get('id', None)
    863 
    864     @property

.../jdaviz/core/template_mixin.py in __getattr__(self, attr)
    108     def __getattr__(self, attr):
    109         if attr[0] == '_' or attr not in self._plugin_traitlets.keys():
--> 110             return super().__getattribute__(attr)
    111 
    112         return getattr(self._plugin, self._plugin_traitlets.get(attr))

.../functools.py in __get__(self, instance, owner)
    967                 val = cache.get(self.attrname, _NOT_FOUND)
    968                 if val is _NOT_FOUND:
--> 969                     val = self.func(instance)
    970                     try:
    971                         cache[self.attrname] = val

.../jdaviz/core/template_mixin.py in selected_item(self)
    283             if not len(items):
    284                 return {}
--> 285             return {k: [item[k] for item in items] for k in items[0].keys()}
    286         return self._get_selected_item(self.selected)
    287 

AttributeError: 'NoneType' object has no attribute 'keys'
```

pllim commented:
For Imviz:

1. Show axis toggle should be removed, as discussed.
2. Layer should only show loaded data, not the Subsets and any loaded markers table. In the screenshot below, it should only show the first 2 entries. Selecting a subset or loaded markers table does not give me any useful options anyway.

![Screenshot 2022-05-26 145103](https://user-images.githubusercontent.com/2090236/170556810-dfe2be7a-4bab-4ca3-9f37-a9220eeeefa6.png)

Otherwise, it seems to work great for Imviz. And I think allowing people to set stretch etc for multiple viewers at once is a killer feature. Great work! 👏

pllim commented:
In Cubeviz, it gets confusing to see FLUX listed twice when select all is done for everything. I understand that one of them is the spectrum but I am not sure if a new user would know that. I wonder if Select All should exclude the spectrum viewer...

![Screenshot 2022-05-26 150225](https://user-images.githubusercontent.com/2090236/170558886-c4af6a9e-c238-434f-8105-e4cc5110e571.png)

p.s. When I am in multi-select mode, sometimes the options like stretch does not show even though I only have one thing selected in that mode. But if I change back to simple mode and select the same thing, it shows. I do not know the exact steps to reproduce this as I was clicking on things to look at other stuff.

pllim commented:
Does it even make sense to allow multi-select for Mosviz? What does this mean?

![Screenshot 2022-05-26 151220](https://user-images.githubusercontent.com/2090236/170560177-decc130d-d436-4bde-be19-3f7b7a514d25.png)

pllim commented:
In Specviz, when I have 3 Subsets and 1 fitted model, I see duplicate entries for Subsets on Select All (screenshot below). Again, like Imviz, I am not sure the value of allowing Subset/Model to be listed in the first place.

![Screenshot 2022-05-26 151700](https://user-images.githubusercontent.com/2090236/170560885-fe28b68c-c45b-46ee-996d-acec75b08068.png)

pllim commented:
Ops, one more thing... dark mode

![Screenshot 2022-05-26 152215](https://user-images.githubusercontent.com/2090236/170562029-42ebc269-4450-4a47-984b-fc029a59d104.png)

pllim commented:
> under "image", it would be good to change "one color per layer" to like "single color" or something similar

This is what @larrybradley requested:

* can we change the colormap mode option from "One color per layer" to "Monochromatic"?

camipacifici commented:
In Specviz:
- I notice that the eye to hide a dataset is not an option anymore. I presume that is because a dataset can be hidden from the data menu?
- However, if I de-select and re-select a dataset in the data menu, some plot options are reset (the color stays, but the line width is back to 1).
- If a subset is created and multiple datasets are present, the subset appears multiple times in the advanced mode.
<img width="388" alt="Screen Shot 2022-05-26 at 9 24 54 PM" src="https://user-images.githubusercontent.com/19998720/170610003-6e87998e-8859-4266-8033-6d31d7b5eb5e.png">
- After that, trying to change the line color triggered a traceback:
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/ipyvue/VueTemplateWidget.py:60, in Events._handle_event(self, _, content, buffers)
     58     getattr(self, "vue_" + event)(data, buffers)
     59 else:
---> 60     getattr(self, "vue_" + event)(data)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/jdaviz/configs/default/plugins/plot_options/plot_options.py:182, in PlotOptions.vue_unmix_state(self, name)
    180 def vue_unmix_state(self, name):
    181     sync_state = getattr(self, name)
--> 182     sync_state.unmix_state()

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/jdaviz/core/template_mixin.py:1509, in PlotOptionsSyncState.unmix_state(self)
   1508 def unmix_state(self):
-> 1509     self._on_value_changed({'new': self.value})
   1510     self.sync = {**self.sync,
   1511                  'mixed': False}

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/jdaviz/core/template_mixin.py:1491, in PlotOptionsSyncState._on_value_changed(self, msg)
   1489         setattr(glue_state, self._glue_name, cmap)
   1490     else:
-> 1491         setattr(glue_state, self._glue_name, msg['new'])
   1492 # need to recompute mixed state
   1493 self._update_mixed_state()

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/echo/core.py:261, in HasCallbackProperties.__setattr__(self, attribute, value)
    260 def __setattr__(self, attribute, value):
--> 261     super(HasCallbackProperties, self).__setattr__(attribute, value)
    262     if self.is_callback_property(attribute):
    263         self._notify_global(**{attribute: value})

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/echo/core.py:76, in CallbackProperty.__set__(self, instance, value)
     74 new = self.__get__(instance)
     75 if old != new:
---> 76     self.notify(instance, old, new)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/utils/matplotlib.py:176, in defer_draw.<locals>.wrapper(*args, **kwargs)
    172 @wraps(func)
    173 def wrapper(*args, **kwargs):
    175     if len(DEFER_DRAW_BACKENDS) == 0:
--> 176         return func(*args, **kwargs)
    178     # Don't recursively defer draws. We just check the first draw_idle
    179     # method since all should be modified in sync.
    180     if isinstance(DEFER_DRAW_BACKENDS[0].draw_idle, DeferredMethod):

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/viewers/matplotlib/state.py:24, in DeferredDrawCallbackProperty.notify(self, *args, **kwargs)
     22 @defer_draw
     23 def notify(self, *args, **kwargs):
---> 24     super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/echo/core.py:125, in CallbackProperty.notify(self, instance, old, new)
    123     return
    124 for cback in self._callbacks.get(instance, []):
--> 125     cback(new)
    126 for cback in self._2arg_callbacks.get(instance, []):
    127     cback(old, new)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/echo/core.py:614, in keep_in_sync.prop2_from_prop1(self, value)
    612 if not self._syncing:
    613     self._syncing = True
--> 614     setattr(self.instance2(), self.prop2, getattr(self.instance1(), self.prop1))
    615     self._syncing = False

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/core/visual.py:201, in VisualAttributes.__setattr__(self, attribute, value)
    197 # if parent has a broadcast method, broadcast the change
    198 if (changed and hasattr(self, 'parent') and
    199     hasattr(self.parent, 'broadcast') and
    200         attribute != 'parent' and not attribute.startswith('_')):
--> 201     self.parent.broadcast('style')

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/core/subset_group.py:170, in SubsetGroup.broadcast(self, item)
    167 @contract(item='string')
    168 def broadcast(self, item):
    169     for s in self.subsets:
--> 170         s.broadcast(item)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/core/subset.py:222, in Subset.broadcast(self, attribute)
    220 if self._broadcasting and self.data.hub:
    221     msg = SubsetUpdateMessage(self, attribute=attribute)
--> 222     self.data.hub.broadcast(msg)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/core/hub.py:215, in Hub.broadcast(self, message)
    213 logging.getLogger(__name__).info("Broadcasting %s", message)
    214 for subscriber, handler in self._find_handlers(message):
--> 215     handler(message)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:70, in SubsetPlugin._on_subset_update(self, *args)
     68 def _on_subset_update(self, *args):
     69     self._sync_selected_from_state(*args)
---> 70     self._get_region_definition(*args)
     71     subset_to_update = self.session.edit_subset_mode.edit_subset[0]
     72     self.subset_select._update_subset(subset_to_update, attribute="type")

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:106, in SubsetPlugin._get_region_definition(self, *args)
    104 def _get_region_definition(self, *args):
    105     self.subset_definition = {}
--> 106     subset_group = [s for s in self.app.data_collection.subset_groups if
    107                     s.label == self.subset_selected][0]
    108     subset_state = subset_group.subset_state
    109     subset_class = subset_state.__class__

IndexError: list index out of range

camipacifici commented:
Apologies for repeating the comment Pey Lian already made about the Subsets. I saw it after I wrote my comments.

camipacifici commented:
The "subset" problem seems more complex than I thought. 
I am in Cubeviz now and I selected 3 subsets in the image viewer. So now I have 4 spectra in the spectrum viewer. Without the eye, I cannot hide them individually. If I de-select the dataset, everything disappears. Maybe I can use the line opacity and set it to zero?
I tried to change the collapse function of the three subsets (in advanced mode), but also the total spectrum changed.
<img width="1617" alt="Screen Shot 2022-05-26 at 9 43 03 PM" src="https://user-images.githubusercontent.com/19998720/170612440-fb6c0f84-f443-4c71-b078-7ed65f85e37c.png">
Trying to change the line opacity of the three subsets (again in advanced mode), I hit this traceback:
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/ipywidgets/widgets/widget.py:756, in Widget._handle_msg(self, msg)
    754         if 'buffer_paths' in data:
    755             _put_buffers(state, data['buffer_paths'], msg['buffers'])
--> 756         self.set_state(state)
    758 # Handle a state request.
    759 elif method == 'request_state':

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/ipywidgets/widgets/widget.py:625, in Widget.set_state(self, sync_data)
    622 if name in self.keys:
    623     from_json = self.trait_metadata(name, 'from_json',
    624                                     self._trait_from_json)
--> 625     self.set_trait(name, from_json(sync_data[name], self))

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/contextlib.py:126, in _GeneratorContextManager.__exit__(self, typ, value, traceback)
    124 if typ is None:
    125     try:
--> 126         next(self.gen)
    127     except StopIteration:
    128         return False

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/traitlets/traitlets.py:1368, in HasTraits.hold_trait_notifications(self)
   1366 for changes in cache.values():
   1367     for change in changes:
-> 1368         self.notify_change(change)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/ipywidgets/widgets/widget.py:686, in Widget.notify_change(self, change)
    683     if name in self.keys and self._should_send_property(name, getattr(self, name)):
    684         # Send new state to front-end
    685         self.send_state(key=name)
--> 686 super(Widget, self).notify_change(change)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/traitlets/traitlets.py:1383, in HasTraits.notify_change(self, change)
   1381 def notify_change(self, change):
   1382     """Notify observers of a change event"""
-> 1383     return self._notify_observers(change)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/traitlets/traitlets.py:1428, in HasTraits._notify_observers(self, event)
   1425 elif isinstance(c, EventHandler) and c.name is not None:
   1426     c = getattr(self, c.name)
-> 1428 c(event)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/jdaviz/core/template_mixin.py:1491, in PlotOptionsSyncState._on_value_changed(self, msg)
   1489         setattr(glue_state, self._glue_name, cmap)
   1490     else:
-> 1491         setattr(glue_state, self._glue_name, msg['new'])
   1492 # need to recompute mixed state
   1493 self._update_mixed_state()

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/echo/core.py:261, in HasCallbackProperties.__setattr__(self, attribute, value)
    260 def __setattr__(self, attribute, value):
--> 261     super(HasCallbackProperties, self).__setattr__(attribute, value)
    262     if self.is_callback_property(attribute):
    263         self._notify_global(**{attribute: value})

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/echo/core.py:76, in CallbackProperty.__set__(self, instance, value)
     74 new = self.__get__(instance)
     75 if old != new:
---> 76     self.notify(instance, old, new)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/utils/matplotlib.py:176, in defer_draw.<locals>.wrapper(*args, **kwargs)
    172 @wraps(func)
    173 def wrapper(*args, **kwargs):
    175     if len(DEFER_DRAW_BACKENDS) == 0:
--> 176         return func(*args, **kwargs)
    178     # Don't recursively defer draws. We just check the first draw_idle
    179     # method since all should be modified in sync.
    180     if isinstance(DEFER_DRAW_BACKENDS[0].draw_idle, DeferredMethod):

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/viewers/matplotlib/state.py:24, in DeferredDrawCallbackProperty.notify(self, *args, **kwargs)
     22 @defer_draw
     23 def notify(self, *args, **kwargs):
---> 24     super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/echo/core.py:125, in CallbackProperty.notify(self, instance, old, new)
    123     return
    124 for cback in self._callbacks.get(instance, []):
--> 125     cback(new)
    126 for cback in self._2arg_callbacks.get(instance, []):
    127     cback(old, new)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/echo/core.py:614, in keep_in_sync.prop2_from_prop1(self, value)
    612 if not self._syncing:
    613     self._syncing = True
--> 614     setattr(self.instance2(), self.prop2, getattr(self.instance1(), self.prop1))
    615     self._syncing = False

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/core/visual.py:201, in VisualAttributes.__setattr__(self, attribute, value)
    197 # if parent has a broadcast method, broadcast the change
    198 if (changed and hasattr(self, 'parent') and
    199     hasattr(self.parent, 'broadcast') and
    200         attribute != 'parent' and not attribute.startswith('_')):
--> 201     self.parent.broadcast('style')

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/core/subset_group.py:170, in SubsetGroup.broadcast(self, item)
    167 @contract(item='string')
    168 def broadcast(self, item):
    169     for s in self.subsets:
--> 170         s.broadcast(item)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/core/subset.py:222, in Subset.broadcast(self, attribute)
    220 if self._broadcasting and self.data.hub:
    221     msg = SubsetUpdateMessage(self, attribute=attribute)
--> 222     self.data.hub.broadcast(msg)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/glue/core/hub.py:215, in Hub.broadcast(self, message)
    213 logging.getLogger(__name__).info("Broadcasting %s", message)
    214 for subscriber, handler in self._find_handlers(message):
--> 215     handler(message)

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:70, in SubsetPlugin._on_subset_update(self, *args)
     68 def _on_subset_update(self, *args):
     69     self._sync_selected_from_state(*args)
---> 70     self._get_region_definition(*args)
     71     subset_to_update = self.session.edit_subset_mode.edit_subset[0]
     72     self.subset_select._update_subset(subset_to_update, attribute="type")

File ~/opt/miniconda3/envs/pr1343/lib/python3.9/site-packages/jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:106, in SubsetPlugin._get_region_definition(self, *args)
    104 def _get_region_definition(self, *args):
    105     self.subset_definition = {}
--> 106     subset_group = [s for s in self.app.data_collection.subset_groups if
    107                     s.label == self.subset_selected][0]
    108     subset_state = subset_group.subset_state
    109     subset_class = subset_state.__class__

IndexError: list index out of range

kecnry commented:
Thanks for all the suggestions!  I tried to distill all of your feedback into remaining action items, but please let me know if you think I'm missing anything you brought up (I'm not intentionally ignoring anything!):

- [ ] Imviz: hide show_axes option
- [ ] add visible "eye" toggle to spectrum lines.
- [ ] dark mode for advanced toggle button
- [ ] re-consider whether layer dropdown should default to first entry instead of empty (the advantage of empty is that it gives easy access to viewer-only options, but can be annoying, especially when the choices reset due to a change in viewer selection or mosviz row).
- [ ] merge subset entries in layer dropdown and link to all applicable states (alternatively provide data prefix as is the case in main)
- [ ] Debug why 2d spectrum layers are not showing image options.  If there aren't any image options, perhaps hide the layers in the layer dropdown.
- [ ] Debug/fix errors reported above with changing line color/opacity of multiple subsets
- [ ] Layer dropdown to exclude markers (or possibly just expose limited options for them?)
- [ ] Test in conjunction with mosviz "freeze properties" functionality


And to consider for the future:
- [ ] Should the data dropdown checkbox just toggle visibility so that plot options don't clear?  Or should it also have an eye button to enable both options?
- [ ] Images can currently be toggled in the image/contour sections... which is below stretch... maybe we want to consider a global eye for images at the top that can toggle both (but then does it have similar mixed state logic or does it "remember" the previous on state?)
- [ ] Can the mosviz "freeze properties" functionality be moved into the plot plugin entirely?  In theory, we could even have per-property freeze abilities now that we control their states.
- [ ] Consider changing the colormap mode option from "One color per layer" to "Monochromatic" (or similar).  We can override this now, but might want a chance for discussion/consensus before we do?  Note that this option is *per-viewer*, and then dictates whether the layers in that viewer have a colormap dropdown _or_ a color selector.  There is also a separate ticket to change the default color to white so it acts more like the default colormap.


And to reply to other comments that I don't think have immediate actions here:
> I switched to the advanced mode, then selected multiple spectra in the 2D viewer, then switched back to the simple mode and the viewer was not responding anymore to changes to the stretch. Actually, the viewer got dimmed like the colorbar was not going all the way to white anymore (in gray scale).

I think I've seen this before on main occasionally (where after a number of interactions glue seems to stop responding to state changes).  I'll see if I can reproduce this more reliably and submit a ticket upstream.  As long as this PR changes the correct state entries, I don't _think_ it was introduced here.

> When I am in multi-select mode, sometimes the options like stretch does not show even though I only have one thing selected in that mode. But if I change back to simple mode and select the same thing, it shows. I do not know the exact steps to reproduce this as I was clicking on things to look at other stuff.

This seems to be different than above since functionality returned when switching back to simple mode, but I have not come across this.  If anyone does see this again, please try to post steps to reproduce and do the following to debug:

```
po = viz.app.get_tray_item_from_name('g-plot-options')
# for stretch (see the init in plot_options.py for the any other attribute name)
print(po.stretch.linked_states, 
         po.stretch.value,
         [getattr(state, po.stretch._glue_name) for state in po.stretch.linked_states])
```

> I tried to change the collapse function of the three subsets (in advanced mode), but also the total spectrum changed.

The collapse function is per-viewer, not per-layer (in glue), so all spectrum will react regardless of the layer selection.  This is where mixing the viewer and layer options can cause some confusion, but you'll notice that it has a "1" next to it which is supposed to tell that it affects the entire spectrum viewer.  In the future, I'm hoping we can make that more clear by using the same number/letters when labeling the viewers.

pllim commented:
> Should the data dropdown checkbox just toggle visibility so that plot options don't clear? 

If I understood you correctly, I think this change might have performance side effects and break a lot of existing code because right now, what is checked in which viewer controls what you get back from the viewer (does that make sense?). Also I don't know if having everything loaded in all the viewers all the time would cause any headache in the backend.

kecnry commented:
I think I've addressed most/all of the previous comments.  I couldn't reproduce some of the tracebacks anymore, but am not sure the exact workflow to reproduce them, so please do let me know if you can still find ways to trigger them.

I think there's still one big open question for this PR (or possibly as a follow-up): **how should options for subsets be categorized**?  Right now, options for subsets in the profile viewer show up under the line subsection (and are toggled with "show line"), whereas subsets in the image viewer are toggled with "show subset" but do NOT have any exposed options for color or opacity.  

<img width="991" alt="Screen Shot 2022-06-02 at 4 09 01 PM" src="https://user-images.githubusercontent.com/877591/171729597-e45e0588-48d0-499f-b6ce-cd438dd13013.png">

Changes to consider:
1. Create a new subsection for subsets so that subsets in image viewers can have their color/opacity controlled there.  Do subset layers in spectral viewers get moved to this section and out of the line section (in which case there would be duplicated options like line width that would not be possible to group with data entries when in advanced mode)?

2. Lump all visibility toggles together ("show line", "show image", "show subset" -> "visible") _except_ for "show contour" which must be separate.  On one hand this might make the terminology less confusing, but it also adds confusion since contour must be separate.

This is all very hard to explain in text or screenshots, the best way to understand is to open cubeviz, create a spatial subset and a spectral subset, and then play with the plot options, especially in advanced mode.

camipacifici commented:
I agree with @pllim. Lets get this in and test with real use cases.

javerbukh reviewed (approved):
I agree with the other reviewers. This is a marked improvement over the previous plot options, even though it still needs a few iterations. I would suggest delaying the 2.7 release for another sprint to give time for any bug fixes after we merge this. Great work!!

P.S. maybe the dark mode fix could be at the top of that bug list? 😄

pllim commented:
How about when Kyle is back next week, we make a list of follow-up issues and then merge this?
